crawl-ref/source/spl-cast.cc

```diff
@@ -3408,6 +3464,330 @@ void do_demonic_magic(int pow, int rank)
+
+        if (!setup_fragmentation_beam(tempbeam, 0, &you,
+            *ri, true, nullptr, temp))
+        {
+            continue;
+        }
+
+        //check to see if you can hit
+        if (grid_distance(*ri, target.target) > tempbeam.ex_size)
+            continue;
+
+        dests.emplace_back(*ri);
+    }
+
+    if (dests.empty())
+        return false;
+    target.target = dests[random2(dests.size())];
+    return true;
+}
+
+static void _find_tabcast_boulder_target(dist &target)
+{
+    coord_def offset = target.target - you.pos();
+    offset.x = max(min(offset.x, 1), -1);
+    offset.y = max(min(offset.y, 1), -1);
+    target.target = you.pos() + offset;
+}
+
+static bool _find_tabcast_paragon_target(dist &target)
+{
+    int radius = 2;
+    const monster* paragon = find_player_paragon();
+    if (paragon && paragon_charge_level(*paragon) == 2)
+        radius = 3;
+
+    for (distance_iterator di(target.target, true, false, radius); di; ++di)
+    {
+        if (actor_at(*di) || !in_bounds(*di)
+            || cell_is_solid(*di) || !you.see_cell(*di)
+            || grid_distance(you.pos(), *di) > calc_spell_range(SPELL_PLATINUM_PARAGON, 0, true, false))
+        {
+            continue;
+        }
+
+        target.target = *di;
+        return true;
+    }
+
+    return false;
+}
+
+static bool _find_tabcast_tempering_target(dist &target)
+{
+    vector<coord_def> dests;
+    for (radius_iterator ri(target.target, 1, C_SQUARE, LOS_SOLID, true); ri; ++ri)
+    {
+        monster* mons = monster_at(*ri);
+        if (!mons || !you.can_see(*mons) || !mons->friendly()
+        || mons->has_ench(ENCH_TEMPERED) || !is_valid_tempering_target(*mons, you))
+        {
+            continue;
+        }
+
+        dests.emplace_back(*ri);
+    }
+
+    if (dests.empty())
+        return false;
+    target.target = dests[random2(dests.size())];
+    return true;
+}
+
+void tabcast_spell(coord_def &pos)
+{
+    const spell_type spell = static_cast<spell_type>(you.attribute[ATTR_TABCAST_SPELL]);
+
+    dist target;
+    target.target = pos;
+    switch (spell)
+    {
+    case SPELL_FULMINANT_PRISM:
+        if (!_find_tabcast_prism_target(target))
+            return;
+        break;
+    case SPELL_LRD:
+        if (!_find_tabcast_lrd_target(target))
+            return;
+        break;
+    case SPELL_GELLS_GAVOTTE:
+    case SPELL_SPLINTERFROST_SHELL:
+    case SPELL_BOULDER:
+        _find_tabcast_boulder_target(target);
+        break;
+    case SPELL_PLATINUM_PARAGON:
+        _find_tabcast_paragon_target(target);
+        break;
+    case SPELL_PERCUSSIVE_TEMPERING:
+        _find_tabcast_tempering_target(target);
+        break;
+    default:
+        break;
+    }
+
+    you.attribute[ATTR_TABCASTING] = 1;
+    cast_a_spell(false, spell, &target);
+    you.attribute[ATTR_TABCASTING] = 0;
+}
+
+static bool _is_channeled_spell(spell_type spell)
+{
+    return spell == SPELL_SEARING_RAY || spell == SPELL_FLAME_WAVE
```

Comment:
Would it make sense to add this as a spflag over in spl-data, so that we can just check that instead of hardcoding the currently-existing channeled spells?